### DISCUSSION TITLE
protection against extreme volume

### DISCUSSION THREAD

**Discussion by jonascant** [pschatzmann/arduino-audio-tools#1634] [Q&A] [answered]
### Problem Description

I am working on a project, that involves very large amplification of very quiet sounds, using the inmp441. I am creating a device, that people will connect headphones to and walk around to catch sometimes extremely quiet sounds. If however they encounter a loud noise for any reason, i need to limit the output through the headphone amplifier (pcm5102). How can the output volume limited?
Thanks!

### Device Description

ESP32 wroom (cheap clone)

### Sketch

```cpp
#include "AudioTools.h"
 
AudioInfo info(96000, 2, 32);
I2SStream in;
I2SStream out; 
VolumeStream volume(out);
ConverterFillLeftAndRight<int32_t> filler(RightIsEmpty); // fill both channels - or change to RightIsEmpty
StreamCopy copier(volume, in); // copies sound into i2s


// Arduino Setup
void setup(void) { 
  // Open Serial 
  Serial.begin(115200);
  // change to Warning to improve the quality
  AudioLogger::instance().begin(Serial, AudioLogger::Warning); 

  // start I2S in
  Serial.println("starting I2S...");
  auto config_in = in.defaultConfig(RX_MODE);
  config_in.copyFrom(info); 
  config_in.i2s_format = I2S_STD_FORMAT;
  config_in.is_master = true;
  // config_in.bits_per_sample = 32;
  config_in.port_no = 0;
  config_in.pin_bck = 14;
  config_in.pin_ws = 27;
  config_in.pin_data = 12;
  // config_in.fixed_mclk = sample_rate * 256
  // config_in.pin_mck = 2
  in.begin(config_in);

  // start I2S out
  auto config_out = out.defaultConfig(TX_MODE);
  config_out.copyFrom(info); 
  config_out.port_no = 1;
  config_out.pin_bck = 26;
  config_out.pin_ws = 25;
  config_out.pin_data = 22;
  out.begin(config_out);

  auto configvol = volume.defaultConfig();
  configvol.allow_boost = true; 
  //configvol.bits_per_sample = 16;
  volume.begin(configvol);
  volume.setVolume(20.0); 

  Serial.println("I2S started...");

}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy(filler);
}
```

### I have checked existing issues, discussions and online documentation

- [X] I confirm I have checked existing issues, discussions and online documentation

**Comment by pschatzmann** [comment, marked as answer]
You can send the output to an [MultiOutput](https://github.com/pschatzmann/arduino-audio-tools/wiki/Splitting-and-Merging-Audio) instead of the VolumeStream.
Add the [VolumeOutput to measure the received volume](https://github.com/pschatzmann/arduino-audio-tools/wiki/Determining-the-Volume) and the VolumeStream to regulate the output.

You can now make the output volume dependent from the input volume....

**Comment by pschatzmann** [comment]
I have renamed the VolumeOutput to VolumeMeter and extended the class, so that it can be used to wrap an input or output class.
```cpp
#include "AudioTools.h"
 
AudioInfo info(96000, 2, 32);
I2SStream in;
VolumeMeter meter(in);
I2SStream out; 
VolumeStream volume(out);
ConverterFillLeftAndRight<int32_t> filler(RightIsEmpty); // fill both channels - or change to RightIsEmpty
StreamCopy copier(volume, meter); // copies sound into i2s
```

This allows you to adjust the volume based on the input w/o any delay...

**Comment by jonascant** [comment]
Awesome, thanks. I failed to configure the meter with meter.defaultConfig. The Compiler says: 
` error: 'class audio_tools::VolumeMeter' has no member named 'defaultConfig'
   48 |   auto cfg_meter = meter.defaultConfig();
`
otherwise, would the following be the way to limit the volume like you suggested? I guess having hard max volume values is not ideal. it should probably be dynamically calculated, right?

`#include "AudioTools.h"
 
AudioInfo info(96000, 2, 32);
I2SStream in;
VolumeMeter meter(in);
I2SStream out; 
VolumeStream volume(out);
ConverterFillLeftAndRight<int32_t> filler(RightIsEmpty); // fill both channels - or change to RightIsEmpty
StreamCopy copier(volume, in); // copies sound into i2s

//volume limiter to avoid extreme loudness
const int save_in_vol = 10; //unit?
const int save_out_vol = 10;
const int max_out_vol = 20; // for amplification of quiet sounds

// Arduino Setup
void setup(void) { 
  // Open Serial 
  Serial.begin(115200);
  // change to Warning to improve the quality
  AudioLogger::instance().begin(Serial, AudioLogger::Warning); 

  // start I2S in
  Serial.println("starting I2S...");
  auto config_in = in.defaultConfig(RX_MODE);
  config_in.copyFrom(info); 
  config_in.i2s_format = I2S_STD_FORMAT;
  config_in.is_master = true;
  // config_in.bits_per_sample = 32;
  config_in.port_no = 0;
  config_in.pin_bck = 14;
  config_in.pin_ws = 27;
  config_in.pin_data = 12;
  // config_in.fixed_mclk = sample_rate * 256
  // config_in.pin_mck = 2
  in.begin(config_in);

  // start I2S out
  auto config_out = out.defaultConfig(TX_MODE);
  config_out.copyFrom(info); 
  config_out.port_no = 1;
  config_out.pin_bck = 26;
  config_out.pin_ws = 25;
  config_out.pin_data = 22;
  out.begin(config_out);

  auto configvol = volume.defaultConfig();
  configvol.allow_boost = true; 
  //configvol.bits_per_sample = 16;
  volume.begin(configvol);

  auto cfg_meter = meter.defaultConfig();
  cfg_meter.copyFrom(info);
  meter.begin(cfg_meter);

  volume.setVolume(20.0); 

  Serial.println("I2S started...");

}

// Arduino loop - copy sound to out 
void loop() {
  Serial.print(meter.volume(0));
  if (meter.volume > save_in_vol){
    Serial.print ("Volume is too loud");
    volume.setVolume(save_out_volume);
  }else{
    volume.setVolume(max_out_volume);
  }
    
  copier.copy(filler);
}`

**Comment by pschatzmann** [comment]
just call meter.begin(info);

I would try a dynamically calculated volume value or you can just switch off the output (setting the volume to 0) if the input is too high.

It should simplify the calculation logic if you configure the VolumeStream to use a linear volume logic.

**Comment by jonascant** [comment]
Calling meter.begin(info) results in this:
`error: no matching function for call to 'audio_tools::VolumeMeter::begin(audio_tools::AudioInfo&)'
   52 |   meter.begin(info);
`

I dont yet understand what a linear VolumeStream logic is, but i will try to get that going!

**Reply by pschatzmann** [reply]
calling setAudioInfo(info) should work, otherwise install the latest corrections

**Comment by jonascant** [comment]
It compiles using meter.begin(); but the volume is always 0.00 and the logger complains:
`0.00[E] AudioStreams.h : 1661 - begin not called!
`

**Comment by jonascant** [comment]
Sorry, i was not on the latest version of your library. It compiles fine now

**Comment by jonascant** [comment]
Something is still not right. The code runs without error, but the volume value is always zero or "inf" when using volumeDB().

`#include "AudioTools.h"
 
AudioInfo info(96000, 2, 32);
I2SStream in;
VolumeMeter meter(in);
I2SStream out; 
VolumeStream volume(out);
ConverterFillLeftAndRight<int32_t> filler(RightIsEmpty); // fill both channels - or change to RightIsEmpty
StreamCopy copier(volume, in); // copies sound into i2s

//volume limiter to avoid extreme loudness
const int save_in_vol = 10; //unit?
const int save_out_vol = 10;
const int max_out_vol = 20; // for amplification of quiet sounds

// Arduino Setup
void setup(void) { 
  // Open Serial 
  Serial.begin(115200);
  // change to Warning to improve the quality
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  //setAudioInfo(info);

  // start I2S in
  Serial.println("starting I2S...");
  auto config_in = in.defaultConfig(RX_MODE);
  config_in.copyFrom(info); 
  config_in.i2s_format = I2S_STD_FORMAT;
  config_in.is_master = true;
  // config_in.bits_per_sample = 32;
  config_in.port_no = 0;
  config_in.pin_bck = 14;
  config_in.pin_ws = 27;
  config_in.pin_data = 12;
  // config_in.fixed_mclk = sample_rate * 256
  // config_in.pin_mck = 2
  in.begin(config_in);

  // start I2S out
  auto config_out = out.defaultConfig(TX_MODE);
  config_out.copyFrom(info); 
  config_out.port_no = 1;
  config_out.pin_bck = 26;
  config_out.pin_ws = 25;
  config_out.pin_data = 22;
  out.begin(config_out);

  auto configvol = volume.defaultConfig();
  configvol.allow_boost = true; 
  //configvol.bits_per_sample = 16;
  volume.begin(configvol);

  meter.begin(info);

  volume.setVolume(20.0); 

  Serial.println("I2S started...");

}

// Arduino loop - copy sound to out 
void loop() {
  Serial.print(meter.volumeDB());
  if (meter.volumeDB() > save_in_vol){
    Serial.print ("Volume is too loud");
    volume.setVolume(save_out_vol);
  }else{
    volume.setVolume(max_out_vol);
  }
    
  copier.copy(filler);
}
`
(how can i get code highlighting here?)

**Reply by pschatzmann** [reply]
db is -inf when there is no volume and 0 for full volume

**Comment by jonascant** [comment]
But how come its constantly 0.00 with a working i2s stream from an inmp442?

**Reply by pschatzmann** [reply]
Not sure: what volume and volume ratio values are you getting ?

**Reply by jonascant** [reply]
I guess none. I don't know what they are. But since i just want to see the volume of the input stream, why do i need to set anything? I don't have control over the input stream. I think i misunderstood something here.

**Reply by pschatzmann** [reply]
You don't need to set anything, but just print the result from volume() and volumeRatio() in the loop.
Volume gives you the max amplitude for each loop step, the ratio gives you values between 0.0 and 1.0

The max possible amplitude for 32 bits is 2,147,483,647

**Reply by jonascant** [reply]
Got it.

  Serial.print(meter.volume());
  Serial.print(" ");
  Serial.print(meter.volumeDB());
  Serial.print(" ");
  Serial.println(meter.volumeRatio());

gives this:

0.00 inf 0.00
0.00 inf 0.00
0.00 inf 0.00
0.00 inf 0.00
0.00 inf 0.00
...

**Reply by jonascant** [reply]
but the stream is working, i hear it through the dac

**Reply by jonascant** [reply]
i get the same output, if i run the meter on the dac with 
`VolumeMeter meter(out);
`

**Reply by jonascant** [reply]
the issue was, i did not use meter as the source in the copier. Now it prints the values! Thanks for your help.
